ext/js/lib/js/require_remote/url_resolver.rb

```diff
@@ -0,0 +1,45 @@
+module JS
+  class RequireRemote
+    ScriptLocation = Data.define(:url, :filename)
+
+    # When require_relative is called within a running Ruby script,
+    # the URL is resolved from a relative file path based on the URL of the running Ruby script.
+    # It uses a stack to store URLs of running Ruby Script.
+    # Push the URL onto the stack before executing the new script.
+    # Then pop it when the script has finished executing.
+    class URLResolver
+      def initialize(base_url)
+        @url_stack = [base_url]
+      end
+
+      def get_location(relative_feature)
+        filename = filename_from(relative_feature)
+        url = resolve(filename)
+        ScriptLocation.new(url, filename)
+      end
+
+      def push(url)
+        @url_stack.push url
+      end
+
+      def pop()
+        @url_stack.pop
+      end
+
+      private
+
+      def filename_from(relative_feature)
+        if relative_feature.end_with?(".rb")
+          relative_feature
+        else
+          "#{relative_feature}.rb"
+        end
+      end
+
+      # Return a URL object of JavaScript.
+      def resolve(relative_filepath)
+        JS.global[:URL].new relative_filepath, @url_stack.last
```

Comment:
I think we need to carefully handle relative path -> absolute path resolution here. IIUC JavaScript's `new URL(path, base)` does not behave as we expect for `Kernel#require_relative` (and also `File.expand_path`).

Here is an example:
| `path, base` | `File.expand_path(path, base)` | `new URL(path, "https://example.com" + base)` |
|:-------------------|:---------------------|:---------------------|
| path="b.rb", base="/lib" | `/lib/b.rb` | `https://example.com/b.rb`
| path="./b.rb", base="/lib" | `/lib/b.rb` | `https://example.com/lib/b.rb`

I think the test suite of `File.expand_path` would be helpful for you to see what should be cared https://github.com/ruby/ruby/blob/master/spec/ruby/core/file/expand_path_spec.rb

Reply:
I believe this behavior of the  JavaScript's URL constructor is as expected.

In most cases, I assume that the first value for `base` is the URL of an HTML file, such as http://exapmle.com/index.html, or a URL such as http://exapmle.com/index without the extension.
Also, once the ruby script is loaded, the `base` value will be the URL of the ruby script, such as http://exapmle.com/a.rb.

So we want the string after the last `/` in the `base` value to be ignored.

Reply:
Ah, ok. I misunderstood the second argument. If the second argument is always a URL to a file, then it makes sense to me.

Reply:
Anyway, unit tests covering this area would be preferable.

Reply:
I wrote unit testsfor JS::RequireRemote::URLResolver.
Please let me know if there are any tests that you think we should add.